code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/GithubTokenHandler.java

```diff
@@ -19,100 +19,123 @@
 
-    private static final DynamicStringProperty GITHUB_CLIENT_ID = ArchaiusUtil.getString("api.auth.github.client.id");
-    private static final DynamicStringProperty GITHUB_CLIENT_SECRET = ArchaiusUtil.getString("api.auth.github.client.secret");
-    private static final DynamicStringProperty GITHUB_USER_INFO_URL = ArchaiusUtil.getString("api.auth.github.user.url");
-
-    public Token getGithubAccessToken(ApiRequest request) throws IOException {
-
-        List<RequestParam> requestData = new ArrayList<>();
-
-        String code = null;
+    public Token getGithubAccessToken(ApiRequest request) {
         Map<String, Object> requestBody = CollectionUtils.toMap(request.getRequestObject());
-        code = (String) requestBody.get(GITHUB_REQUEST_TOKEN);
-
-        requestData.add(new RequestParam(CLIENT_ID, GITHUB_CLIENT_ID.get()));
-        requestData.add(new RequestParam(CLIENT_SECRET, GITHUB_CLIENT_SECRET.get()));
-        requestData.add(new RequestParam(GITHUB_REQUEST_TOKEN, code));
+        String code = ObjectUtils.toString(requestBody.get(GITHUB_REQUEST_TOKEN));
+        Map<String, Object> jsonData = client.getAccessToken(code);
+        String token = (String) jsonData.get("access_token");
+        List<String> idList = new ArrayList<>();
+        List<String> orgNames = new ArrayList<>();
 
-        Map<String, Object> jsonData = null;
+        GithubAccountInfo userAccountInfo = client.getUserAccountInfo(token);
+        List<GithubAccountInfo> orgAccountInfo = client.getOrgAccountInfo(token);
 
-        HttpResponse response = Request.Post(GITHUB_URL.get()).addHeader("Accept", "application/json").bodyForm(requestData).execute().returnResponse();
-        int statusCode = response.getStatusLine().getStatusCode();
-        if (statusCode != 200) {
-            throw new ClientVisibleException(statusCode);
+        if (StringUtils.isNotEmpty(userAccountInfo.getAccountId())) {
+            idList.add(userAccountInfo.getAccountId());
         }
-        jsonData = jsonMapper.readValue(response.getEntity().getContent());
 
-        if (jsonData.get("error") != null) {
-            throw new ClientVisibleException(ResponseCodes.BAD_REQUEST, (String) jsonData.get("error_description"));
+        for (GithubAccountInfo info : orgAccountInfo) {
+            if (StringUtils.isNotEmpty(info.getAccountId())) {
+                idList.add(info.getAccountId());
+                orgNames.add(info.getAccountName());
+            }
         }
 
-        AccountInfo accountInfo = getAccountInfo((String) jsonData.get("access_token"));
-        if (StringUtils.isNotEmpty(accountInfo.accountId)) {
-            jsonData.put("account_id", accountInfo.accountId);
+        if (SECURITY.get()) {
+            if (null == getWhitelistedUser(idList)) {
+                throw new ClientVisibleException(ResponseCodes.UNAUTHORIZED, "User is not whitelisted");
```

Comment:
This is the wrong use of this constructor.  The error code should be something like "UserNotWhitelisted", and then "User is not..." is the message.  You are setting the message for the code.  Use the constructor with 4 args and do (status, code, message, detail) where detail will be null in your situation.

Having said all that, typically a 401 does not tell you why you have been reject.  So I would just go with the 1 arg constructor `new ClientVisibleException(ResponseCodes.UNAUTHORIZED)`